Changing API to encompas multiple output bits.

RolfHuisman opened the pull request:
Was debugging a bit, because I thought it was a small issue, but seems the amount of output bits was set to the address length and al output bits were set, which resulted in all kinds of situations. So rewrote the api to support different address lengths from the output sizes, taking the bit array size as the value size.

Commit:
Changing API to encopas multiple output bits.

crazy4pi314 commented on samples/implicitQRAM/Program.qs:
```diff
@@ -41,12 +42,12 @@
-        return [five, four, one];
+    //Generates binary representation of [5,4,1]=>[3,2,1]
+    function GenerateMemoryData() : (Bool[],Bool[])[] {
+        let fiveGivesThree = (IntAsBoolArray(5,3),IntAsBoolArray(3,2));
```
If I understand the naming here correctly, this means that at address number 5, the value 3 is stored?  You used `Has` later in the tests and that might be more consistent?

crazy4pi314 commented on src/Library.qs:
```diff
@@ -19,63 +19,59 @@
+    function ImplicitQRAMOracle(DataValues : (Bool[],Bool[])[]) : QRAM {
+        mutable addressSize = 0;
+        mutable valueSize = 0;
+        for((_,(address,value)) in Enumerated(DataValues)){
```
Since you are dropping/not using the index generated with `Enumerated` then you can just do
```suggestion
        for ((address, value)  in DataValues){
```

crazy4pi314 commented on src/Library.qs:
```diff
@@ -19,63 +19,59 @@
+            w/ DataSize <- valueSize;
+    }
+    
+    internal function SingleImplicitQRAMOracle(DataValue : (Bool[],Bool[])) : SingleBitQRAM {
```
It looks like here that you don't use `DataValue` directly and immediately unwrap it on 44 so you could just name the arguments directly:
```suggestion
    internal function SingleImplicitQRAMOracle(address : Bool[], value : Bool[]) : SingleBitQRAM {
```
This is the same as `SingleImplicitQRAMOracle((address : Bool[], value : Bool[]))` due to singleton tuple equivalence (which I read about a while back [here](https://docs.microsoft.com/en-us/quantum/user-guide/language/types#singleton-tuple-equivalence) ).

crazy4pi314 commented on src/Library.qs:
```diff
@@ -19,63 +19,59 @@
+    }
+    
+    internal function SingleImplicitQRAMOracle(DataValue : (Bool[],Bool[])) : SingleBitQRAM {
+        let (address,value) = DataValue;
```
If taken with above suggestion, this line is not needed.
```suggestion

```

crazy4pi314 commented on src/Library.qs:
```diff
@@ -19,63 +19,59 @@
     //Internal QRam type for a single QRAM Bit (is composed to form a full memory block)
-    internal newtype SingleBitQRAM = (Lookup : ((Qubit[], Qubit) => Unit is Adj + Ctl), 
-        AddressSize : Int);
+    internal newtype SingleBitQRAM = (Lookup : ((Qubit[], Qubit[]) => Unit is Adj + Ctl));
```
This really isn't a single bit qRAM anymore, any ideas for a better name?

glassnotes replied:
Since this is the most general case (multiple address bits in, and multiple value bits out) I half-seriously suggest LooqUpTable / LooqUpBit, or just the general qROM.

crazy4pi314 replied:
#sorrynotsorry but yeah let's keep this discussion open, I'll file it as an issue.

crazy4pi314 reviewed (approved):
I think this is looking pretty great, thanks for the help! I left a few comments and I think once we get this merged I got some new ideas to work with based on what you did here 😄

crazy4pi314 committed:
Apply suggestions from code review

Just committing these after talking to @RolfHuisman, want to get master tests working!
